src/util/indexing.jl

```diff
@@ -0,0 +1,263 @@
+Base.(:(==))(x::Indexer, y::Indexer) = isequal(x, y)
+Base.haskey(x::Indexer, key::Symbol) = haskey(x.lookup, key)
+Base.haskey{S <: AbstractString}(x::Indexer, key::S) = haskey(x, convert(Symbol, key))
+Base.haskey(x::Indexer, key::Real) = 1 <= key <= length(x.names)
+Base.keys(x::Indexer) = names(x)
+
+
+"Completely replace the names in an Indexer with a new Vector of Symbols."
+function names!(x::Indexer, names::Vector{Symbol})
+    if length(names) != length(x)
+        throw(ArgumentError("Length of new names doesn't match length of Index."))
+    end
+    names = make_unique(names)
+    if length(x.names) != length(x.lookup)
+        throw(ArgumentError("Index corrupted by duplicate symbols in names."))
+    end
+    rename!(x, x.names, names)
+    return x
+end
+
+
+"Completely replace the names in an Indexer with a new Vector of Strings."
+function names!{S <: AbstractString}(x::Indexer, names::Vector{S})
+    names!(x, convert(Vector{Symbol}, names))
+end
+
+
+function rename!(x::Indexer, names)
+    for (from, to) in names
+        if !haskey(x, from)
+            throw(ArgumentError("Cannot rename $from to $to, $from does not exist in the Indexer."))
+        end
+        if haskey(x, to)
+            throw(ArgumentError("Cannot rename $from to $to, $to already exists in the Indexer."))
+        end
+        # Change the name array
+        nameidx = findfirst(x.names, from)
+        x.names[nameidx] = to
+        # Change the dictionary
+        x.lookup[to] = val = pop!(x.lookup, from)
+    end
+    return x
+end
+
+
+"""
+Rename a set of names in an Indexer, to a new set of names.
+
+This will rename such that the name in from[1] will be renamed to to[1],
+from[2] will be renamed to to[2] and so on.
+"""
+function rename!(x::Indexer, from::Vector{Symbol}, to::Vector{Symbol})
+    return rename!(x, zip(from, to))
+end
+
+
+"""
+Rename a set of names in an Indexer, to a new set of names.
+
+This will rename such that the name in from[1] will be renamed to to[1],
+from[2] will be renamed to to[2] and so on.
+"""
+function rename!{S <: AbstractString}(x::Indexer, from::Vector{S}, to::Vector{S})
+    return rename!(x, convert(Vector{Symbol}, from), convert(Vector{Symbol}, to))
+end
+
+
+"Rename a single name (from) in Indexer x, to a new name (to)."
+rename!(x::Indexer, from::Symbol, to::Symbol) = rename!(x, ((from, to),))
+
+"Rename a single name (from) in Indexer x, to a new name (to)."
+function rename!{S <: AbstractString}(x::Indexer, from::S, to::S)
+    return rename!(x, ((convert(Symbol, from), convert(Symbol, to)),))
+end
+
+
+"""
+Rename a indexer according to some function, f, that accepts one argument.
+
+Each name in the Indexer x, will be input to, f, and the output will be
+what that name will be renamed to.
+
+Therefore the function, f, will need to return Symbols or Strings.
+"""
+rename!(x::Indexer, f::Function) = rename!(x, [(x,f(x)) for x in x.names])
```

Comment:
I think `rename!` should take a function as the first argument for consistency with other standard functions: `map`, `foldl`, `foldr`, etc.